### PR TITLE
refactor(select): Move adapter logic into foundation

### PR DESCRIPTION
**Pull request by bonniezhou**
- Proxy to foundation for `get/set required`
- Move `checkValidity` and `setValid` adapter logic to foundation
- Simplify `openMenu` in adapter
- Remove unused adapter method `getSelectedMenuItem`
- Simplify `toggleClassAtIndex` adapter method into `addClassAtIndex` and `removeClassAtIndex`

### CODE REVIEW

**Comment by bonniezhou** [comment]
@abhiomkar this PR is incomplete but I wanted to get an initial review from you to see if these are the changes we want to make. 
In some cases, moving more adapter logic into the foundation may make it more complicated to wrap in other frameworks. For example, I'm guessing that `checkValidity` became an adapter method for wrapping in MDC React, because React [uses props instead of classes](https://github.com/material-components/material-components-web-react/blob/master/packages/select/index.tsx#L219).

File: packages/mdc-select/foundation.ts [hunk 1]
```diff
@@ -307,11 +318,33 @@ export class MDCSelectFoundation extends MDCFoundation<MDCSelectAdapter> {
...
+      return this.selectedIndex_ !== -1 && (this.selectedIndex_ !== 0 || Boolean(this.adapter_.getValue()));
+    }
+    return true;
+  }
+
+  setRequired(isRequired: boolean) {
+    if (isRequired) {
+      this.adapter_.setSelectedTextAttr('aria-required', isRequired.toString());
+    } else {
+      this.adapter_.removeSelectedTextAttr('aria-required');
```
Comment (abhiomkar) [inline packages/mdc-select/foundation.ts]: We can use `setSelectedTextAttr` here, setting attr value to `false`.

File: packages/mdc-select/component.ts [hunk 2]
```diff
@@ -348,29 +344,14 @@ export class MDCSelect extends MDCComponent<MDCSelectFoundation> {
...
-        }
-      },
       setSelectedText: (text: string) => this.selectedText_.textContent = text,
+      getSelectedTextAttr: (attr: string) => this.selectedText_.getAttribute(attr),
+      setSelectedTextAttr: (attr: string, value: string) => this.selectedText_.setAttribute(attr, value),
+      removeSelectedTextAttr: (attr: string) => this.selectedText_.removeAttribute(attr),
       openMenu: () => {
         if (!this.menu_.open) {
           this.menu_.open = true;
           this.isMenuOpen_ = true;
```
Comment (abhiomkar) [inline packages/mdc-select/component.ts]: `this.isMenuOpen_` - We may not need to store this state in component,  we can use `this.menu_.open` instead.

File: packages/mdc-select/component.ts [hunk 3]
```diff
@@ -348,29 +344,14 @@ export class MDCSelect extends MDCComponent<MDCSelectFoundation> {
...
-        } else {
-          this.root_.classList.add(cssClasses.INVALID);
-        }
-      },
       setSelectedText: (text: string) => this.selectedText_.textContent = text,
+      getSelectedTextAttr: (attr: string) => this.selectedText_.getAttribute(attr),
+      setSelectedTextAttr: (attr: string, value: string) => this.selectedText_.setAttribute(attr, value),
+      removeSelectedTextAttr: (attr: string) => this.selectedText_.removeAttribute(attr),
       openMenu: () => {
         if (!this.menu_.open) {
```
Comment (abhiomkar) [inline packages/mdc-select/component.ts]: Component should not check for correctness. For example, `if (!this.menu_.open) ...` it should be moved to foundation or it is not required.

File: packages/mdc-select/component.ts [hunk 4]
```diff
@@ -399,17 +379,19 @@ export class MDCSelect extends MDCComponent<MDCSelectFoundation> {
         const menuItem = this.menu_.items[index];
         return menuItem && menuItem.textContent ? menuItem.textContent : '';
       },
-      toggleClassAtIndex: (index: number, className: string, toggle: boolean) => {
+      addClassAtIndex: (index: number, className: string) => {
         const menuItem = this.menu_.items[index];
```
Comment (abhiomkar) [inline packages/mdc-select/component.ts]: Foundation should guarantee to send valid index, again component does not need to check for correctness of input to adapter. ie, `menu_.items[index]` always returns valid menu item.

File: packages/mdc-select/component.ts [hunk 5]
```diff
@@ -348,29 +344,14 @@ export class MDCSelect extends MDCComponent<MDCSelectFoundation> {
...
-      setValid: (isValid: boolean) => {
-        this.selectedText_.setAttribute('aria-invalid', (!isValid).toString());
-        if (isValid) {
-          this.root_.classList.remove(cssClasses.INVALID);
-        } else {
-          this.root_.classList.add(cssClasses.INVALID);
-        }
-      },
       setSelectedText: (text: string) => this.selectedText_.textContent = text,
+      getSelectedTextAttr: (attr: string) => this.selectedText_.getAttribute(attr),
```
Comment (abhiomkar) [inline packages/mdc-select/component.ts]: `getValue` need clean-up too. But we can do it in separate PR.

File: packages/mdc-select/component.ts [hunk 6]
```diff
@@ -281,18 +281,14 @@ export class MDCSelect extends MDCComponent<MDCSelectFoundation> {
    * Sets the control to the required state.
    */
   set required(isRequired: boolean) {
-    if (isRequired) {
```
Comment (abhiomkar) [inline packages/mdc-select/component.ts]: Focus management in `menuSetup_` going to be refactored in later PRs?

**Comment by abhiomkar** [comment]
> @abhiomkar this PR is incomplete but I wanted to get an initial review from you to see if these are the changes we want to make.
> In some cases, moving more adapter logic into the foundation may make it more complicated to wrap in other frameworks. For example, I'm guessing that `checkValidity` became an adapter method for wrapping in MDC React, because React [uses props instead of classes](https://github.com/material-components/material-components-web-react/blob/master/packages/select/index.tsx#L219) [](/material-components/material-components-web-react/blob/HEAD@{2019-07-31T22:09:42Z}/packages/select/index.tsx#L219).

Good question! References to `props` can be delegated to foundation via adapter. (For example, adapter_.isDisabled) which you've done already.
